src/viam/sdk/common/exception.hpp

```diff
@@ -0,0 +1,66 @@
+/// @file common/exception.hpp
+///
+/// @brief Defines custom exceptions for the SDK.
+#pragma once
+#include <grpcpp/support/status.h>
+#include <stdexcept>
+#include <string>
+
+#include <viam/sdk/resource/resource_api.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Exception Classes related to SDK exceptions.
+
+/// @class ErrorCondition
+/// @brief Defines a set of a error conditions to be used in conjunction with
+/// Exception.
+/// @ingroup Exception
+enum class ErrorCondition : uint8_t {
+    k_general = 0,             // Default condition
+    k_connection,              // Issue during connection establishment
+    k_duplicate_registration,  // API or API/Model pair has already been registered
+    k_duplicate_resource,      // Resource has already been added
+    k_grpc,                    // gRPC error from remote machine
+    k_not_supported,           // Behavior not supported by the SDK
+    k_resource_not_found       // Resource does not exist
+};
+
+std::error_condition make_error_condition(ErrorCondition e);
+
+/// @class Exception
+/// @brief Defines an exception type for the SDK.
+/// @ingroup Exception
+class Exception : public std::runtime_error {
+   public:
+    explicit Exception(ErrorCondition condition, const std::string& what);
+    explicit Exception(const std::string& what);
+    virtual ~Exception();
+
+    std::error_condition condition() const noexcept;
```

Comment:
This could return by `const&` too. Did you have a reason for the value return?

Reply:
Ah no reason; returned by const reference instead.